Sources/PackageLoading/ModuleMapGenerator.swift

```diff
@@ -134,7 +137,18 @@ public struct ModuleMapGenerator {
         }
         diagnoseInvalidUmbrellaHeader(includeDir.appending(component: target.c99name))
 
-        try createModuleMap(inDir: wd, type: .directory(includeDir))
+        // Otherwise, if 'include' contains only header files and no subdirectories, use it as the umbrella directory (this is case 3 at https://github.com/apple/swift-package-manager/blob/master/Documentation/Usage.md#creating-c-language-targets).
+        if files.count == walked.count {
+            try createModuleMap(inDir: wd, type: .directory(includeDir))
+            return
+        }
+        
+        // Otherwise, the target's public headers are considered to be incompatible with modules.  Other C targets can still import them, but Swift won't be able to see them.  This is documented as an error, but because SwiftPM has previously allowed it (creating module maps that then cause errors when used), we instead emit a warning.
+        warningStream <<< "warning: the include directory of target '\(target.name)' has "
+        warningStream <<< "a layout that is incompatible with modules; no module map will "
+        warningStream <<< "be generated; consider adding a custom module map to the target"
+        warningStream.flush()
+        try createModuleMap(inDir: wd, type: .empty)
```

Comment:
Should we consider making this warning conditional on the used languages and/or the root package? Presumably, if I am building an all-C package graph, I might not care about modules. That has the danger of shipping a C-only package that is not compatible with Swift, so it might make sense to still emit it for the root package.

Reply:
Since the documentation states that it is an error for packages to not have their headers in a module compatible layout, it seems prudent to at least emit the warning so that package authors can add a custom module map.

Reply:
I think that depends on what we want the future behaviour to be. It seems like we never implemented the error and now non-modular packages exist. Is this considered just something we support for backwards-compatibility and this would become an error in the future? If not, I think there needs be some way to not get the warning.

Reply:
Absolutely, agreed, there needs to be a way to not get the warning, but I think that way would be to add a custom module map to the target.  The warning here is phrased similarly to the other warnings this function can emit, and in each case there should be wording to indicate how to avoid the warning.  The problem now is that packages just fail to build and there's no warning about why.

Reply:
I would not, for example, want to add another bool to the Target() declaration to say whether or not a module is expected.